flink-connector-opensearch/src/test/java/org/apache/flink/streaming/connectors/opensearch/OpensearchSinkTest.java

```diff
@@ -0,0 +1,562 @@
+
+        // Let the whole bulk request fail
+        responses.add(response -> response.setStatusCode(500));
+
+        // setup the next bulk request, and let bulk request succeed
+        testHarness.processElement(new StreamRecord<>("msg-1"));
+
+        // Await for flush to be complete
+        awaitForFlushToFinish(1);
+
+        // setup the requests to be flushed in the snapshot
+        testHarness.processElement(new StreamRecord<>("msg-2"));
+        testHarness.processElement(new StreamRecord<>("msg-3"));
+
+        CheckedThread snapshotThread =
+                new CheckedThread() {
+                    @Override
+                    public void go() throws Exception {
+                        testHarness.snapshot(1L, 1000L);
+                    }
+                };
+        snapshotThread.start();
+
+        // Await for flush to be complete
+        awaitForFlushToFinish(0);
+
+        assertThatThrownBy(snapshotThread::sync)
+                .getCause()
+                .getCause()
+                .hasMessageContaining("Unable to parse response body");
+    }
+
+    /**
+     * Tests that the sink correctly waits for pending requests (including re-added requests) on
+     * checkpoints; we set a timeout because the test will not finish if the logic is broken.
+     */
+    @Test
+    @Timeout(5)
+    public void testAtLeastOnceSink() throws Throwable {
+        final OpensearchSink.Builder<String> builder =
+                new OpensearchSink.Builder<>(
+                        Arrays.asList(new HttpHost("localhost", server.getLocalPort())),
+                        new SimpleSinkFunction<String>());
+        builder.setBulkFlushInterval(500);
+        builder.setFailureHandler(
+                new DummyRetryFailureHandler()); // use a failure handler that simply
+        // re-adds requests
+
+        final OpensearchSink<String> sink = builder.build();
+        final OneInputStreamOperatorTestHarness<String, Object> testHarness =
+                new OneInputStreamOperatorTestHarness<>(new StreamSink<>(sink));
+
+        testHarness.open();
+
+        // setup the next bulk request, and its mock item failures;
+        // it contains 1 request, which will fail and re-added to the next bulk request
+        responses.add(
+                createResponse(
+                        new BulkItemResponse(
+                                1,
+                                OpType.INDEX,
+                                new Failure(
+                                        "test",
+                                        "_doc",
+                                        "1",
+                                        new Exception("artificial failure for record")))));
+
+        testHarness.processElement(new StreamRecord<>("msg"));
+
+        // current number of pending request should be 1 due to the re-add
+        assertThat(sink.getNumPendingRequests()).isEqualTo(1);
```

Comment:
This assertion sometimes fails locally when the test is run in a loop.
```
expected: 1L
 but was: 0L
```

Reply:
Aha, I think I know why, will fix that